### ISSUE TITLE
Azure B2C support

### ISSUE THREAD

**Issue by hugeps** [SocialiteProviders/Providers#418]
Guys,

Which plugin will work with Azure B2C? Looking at providers here all seem to use the URLs from Azure AD, but don't seem to work with AD B2C...

Anyone had a look at this before?

**Comment by johnwc** [comment]
I wrote this provider for our own use, I am more than happy to share with everyone. What is the process of getting it in the list of providers?

**Comment by atymic** [comment]
Hey John, how much does it differ from the normal azure provider? Could the be combined?
Could you PR what you have and i'll have a look, thanks!

**Comment by johnwc** [comment]
It's a bit different, it requires to configure the user flow you want to use for login in the url. I will clean it up and publish it on github. Should I just fork this repo and add it to it?

**Comment by nicosalvadore** [comment]
Hi @johnwc !
I was thinking of using Azure B2C service soon for a work project.
Have you been able to work on this PR yet ?
TIA :)

**Comment by johnwc** [comment]
@nicosalvadore Yes, I need to clean it up a bit and I can then do a PR. We've been using our version since Dec and it's been great so far.

**Comment by fujie** [comment]
Hi, I created PR #733 to add provider for Azure AD B2C.

**Comment by johnwc** [comment]
The provider we wrote is up. You may want to merge ours, as it has more features and actually verifies the token is valid. Also, I would not have the provider cache anything, it shouldn't need this anyways. Have the site's developer cache whatever they need to.
https://github.com/johnwc/Providers/tree/master/src/AzureB2C

**Comment by fujie** [comment]
@johnwc Thank you for let me know about your implementation. It's great!

But I think that caching an OP-metadata and public keys is quite natural to improve performance. And my implementation also validate an id_token by using firebase JWT library.

And I thought that there are unnecessary options there, could please let me know your opinion?
- offline_access is not required since Azure AD B2C does not send refresh_token  by default and if you want to get refresh_token, you have to put another scope to get access_token.
- there is implementation which set 'defaultNonce' as nonce and it is quite un-secure if you want to use the implicit or the hybrid flow of OpenID Connect.
- you set 'id_token' as response_type and 'authorization_code' as grant_type. But by setting 'id_token' as a response_type, Azure AD B2C kicks the implicit flow so the authorization code won't be issued to the client.
- for logout, it is better for site implementers to be able to set 'post_logout_redirect_uri'.

**Comment by fujie** [comment]
Anyway, I think it is fair that adding comments on the pull request to improve the Socialite Provider collection itself. @johnwc could you please add your code as comments on the pull request?

**Comment by fujie** [comment]
@atymic could you please review my PR?

**Comment by atymic** [comment]
Reviewed :)

**Comment by johnwc** [comment]
@fujie 

I think you need to realize that these providers will be used by many, each having their own requirements. Some requiring the need for oAuth tokens. So this provider needs to be able to support being used as OpenID or oAuth, what ever the developing application is requiring. No one is saying that solely setting the scope for offline_access will make B2C return a refresh_token. It is expected that whoever does need the refresh_token will read up on B2C docs to know all what they need to do on the B2C side as well. I know the other claims are not default returned, but that still does not mean that you should not allow other claims to be returned. You're not giving the developer a way to tell it to return more than your two listed. There should be a config setting that allows the developer to tell the provider what all other claims to return.

**Comment by fujie** [comment]
@johnwc Your point of view is true, I'm always appreciate for your comments. So, as I mentioned before, could you please create pull request on my repo?

- For difference between "OAuth Authentication" and OpenID Connect, it is quite dangerous for anyone situation today because only a few people understand correctly. As a co-chair of OpenID Foundation WG, I think I have to inform people which is the right and secure way to "Authenticate" users. So I don't want to implement OAuth related configuration in default without any explanation.
- For additional claims, I already added some instruction to get other claims from Azure AD B2C.

**Comment by lsmith77** [comment]
any advice on where to get the relevant configuration settings from?

**Comment by johnwc** [comment]
@lsmith77 these settings come from the B2C that is setup in your azure subscription. Part of it is from the user flows, and part is from a registered app you create for your web app.

[Tutorial: Create user flows and custom policies in Azure Active Directory B2C](https://docs.microsoft.com/en-us/azure/active-directory-b2c/tutorial-create-user-flows?pivots=b2c-custom-policy)
[Tutorial: Register a web application in Azure Active Directory B2C](https://docs.microsoft.com/en-us/azure/active-directory-b2c/tutorial-register-applications?tabs=app-reg-ga)

**Comment by lsmith77** [comment]
thx for those links ..

what I was hoping for was a bit more detail (like I assume I want "Secret" and not "RSA" for the key policy) but I will try to work through this and then send a PR to update the readme with more instructions as I didn't find a blog post with such details.

**Comment by lsmith77** [comment]
did anyone by chance work on a bicep/terraform/hashi cdk/pulumi setup template?
